### ISSUE TITLE
Some way to allow for Binder branding w/in notebooks etc

### ISSUE THREAD

**Issue by choldgraf** [jupyterhub/binderhub#260]
This will be tricky as we support more and more UIs, but I wonder if it'd be good to have some kind of visual cue within Binder sessions to tell people that they're working on Binder. Maybe a favicon? Or alternatively we had discussed making a configurable logo that's displayed in, e.g., the notebook. This might also be something that organizations want if they start deploying their own BinderHubs.

**Comment by choldgraf** [comment]
another ping on this one since it came up a few times over the last few weeks, in particular @enolfc and @gouarin and maybe @nthiery who mentioned this at the orsay meeting

**Comment by betatim** [comment]
We now have support for appending to a `Dockerfile` (see #459). This means that we have the infrastructure for making this happen inside the live binder. We can inject our own themes for the single user notebook that include some binder branding/advice/link to original repo/GA tracking

Has someone tried modifying the look of the singleuser notebook server already?

**Comment by bitnik** [comment]
For numbers 1 and 2, do you think is it a good idea to enable having custom/extended templates similar to https://github.com/jupyterhub/jupyterhub/pull/1601? If you think so, I can try to bring the same feature to BinderHub.

**Comment by choldgraf** [comment]
Good question - @minrk do you have an intuition for whether we'd be well-off to make templates for BinderHub pages?

**Comment by minrk** [comment]
Yes, I think we can and should. We're already using jinja, so it's mostly a matter of exposing the right configuration, and good use of blocks in our templates, to make it easy to customize the right stuff.

**Comment by bitnik** [comment]
@minrk I made a PR for this.

@betatim I didn't really change the look of single user server but while trying out build appendix feature, I added a button by using `custom.js`. This button (next to logout button) copies sharable user server link (with token) into clipboard. An example repo built after this feature is https://notebooks.gesis.org/binder/v2/gh/binder-examples/requirements/master.

**Comment by bitnik** [comment]
@choldgraf I had myself some difficulties about sharing binder links, so that was the first thing I tried when I saw `build.appendix` is implemented :)

The code is [here](https://github.com/gesiscss/orc/tree/master/binderhub/appendix).

[`run-appendix`](https://github.com/gesiscss/orc/blob/master/binderhub/appendix/run-appendix#L6-L11) script adds the content of [`static/custom.js`](https://github.com/gesiscss/orc/blob/master/binderhub/appendix/static/custom.js) into `~/.jupyter/custom/custom.js` inside user notebook server:
```bash
# add custom js
if [ ! -d ~/.jupyter/custom ]
then
    mkdir -p ~/.jupyter/custom
fi
cat "${appendix_dir}/static/custom.js" >> ~/.jupyter/custom/custom.js
```
And [this `~/.jupyter/custom/custom.js` is executed, when notebook app starts](http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/JavaScript%20Notebook%20Extensions.html#custom.js).

What `static/custom.js` does is basicly that it adds a button next to logout and every time when that button is clicked, it copies the url of jupyter logo (on top left, it actually holds url to share binder session) into clipboard.

I think we could add repo and build links in the same way, since they are already [hold as environment variables](https://github.com/jupyterhub/mybinder.org-deploy/blob/master/mybinder/values.yaml#L119-L120).

**Comment by bitnik** [comment]
Btw I just updated the code to add buttons for repo and build links.

**Comment by jhamman** [comment]
I took a hack at using the approach in #545 today. The templates functionality works great. I wasn't able to hook up the static dependencies (logos, etc). A few notes on what I did as it may be useful to others:

binder deployment: http://binder.pangeo.io/

- new repo holding templates/static resources: https://github.com/pangeo-data/pangeo-custom-binderhub-templates/tree/aef402d181aafb84502edd4c8246de41faeb5a39
- Updated values section in binderhub helm chart: https://github.com/pangeo-data/pangeo-binder/blob/b37e99a700689a1f784ef56882c59a8ea382f61a/deploy/binder_config.yml#L14-L40

Like I said, this worked pretty well for the templates but it seems the static resources are clearly handled differently. I think this feature will be used fairly often by new deployments so I really encourage working up some proper documentation. I'm curious if there are any known uses of this PR and if they can be shared (@bitnik - I image you had a use case for the features in #545).

**Comment by bitnik** [comment]
@jhamman [your url prefix for static files](https://github.com/pangeo-data/pangeo-binder/blob/b37e99a700689a1f784ef56882c59a8ea382f61a/deploy/binder_config.yml#L18) is the same as [default url prefix](https://github.com/jupyterhub/binderhub/blob/master/binderhub/app.py#L479) and it is overridden by it. You have to set it different than it, e.g. `extra_static/`. Then you have to change url of your static files in templates, in this case for your logo: `src="/extra_static/pangeo.svg"`

I think even better you can define a template variable and use it in templates as url prefix for extra static files:
```yaml
template:
  path: "/etc/binderhub/templates/templates/templates"
  static:
    path: "/etc/binderhub/templates/templates/templates"
    urlPrefix: extra_static/
  variables:
    EXTRA_STATIC_URL_PREFIX: "<binderhub_base_url>extra_static/"
```
and then you can use: `src="{{ EXTRA_STATIC_URL_PREFIX }}pangeo.svg"`

I hope this works. And you are right, i will try to document it soon!

**Comment by jhamman** [comment]
Thanks @bitnik - this is up and working now (thanks also to @dsludwig).
